src/legacy/core_plugins/data/public/filter/filter_manager/filter_manager.ts

```diff
@@ -0,0 +1,245 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { Filter, isFilterPinned, FilterStateStore } from '@kbn/es-query';
+
+import _ from 'lodash';
+import { Subject, Subscription } from 'rxjs';
+
+import { npSetup } from 'ui/new_platform';
+
+// @ts-ignore
+import { compareFilters } from './lib/compare_filters';
+// @ts-ignore
+import { mapAndFlattenFilters } from './lib/map_and_flatten_filters';
+// @ts-ignore
+import { uniqFilters } from './lib/uniq_filters';
+// @ts-ignore
+import { extractTimeFilter } from './lib/extract_time_filter';
+// @ts-ignore
+import { changeTimeFilter } from './lib/change_time_filter';
+
+import { PartitionedFilters } from './partitioned_filters';
+import { FilterStateManager } from './filter_state_manager';
+
+// const COMPARATOR_OPTIONS = { negate: true, disabled: true };
+
+export class FilterManager {
+  filterState: FilterStateManager;
+  indexPatterns: any;
+  filters: Filter[] = [];
+  updated$: Subject<any> = new Subject();
+  fetch$: Subject<any> = new Subject();
+  updateSubscription$: Subscription | undefined;
+
+  constructor(indexPatterns: any, filterState: FilterStateManager) {
```

Comment:
i would prefer to see this build the other way arround, instead of passing FilterStateManager to FilterManager i would pass FilterManager to FilterStateManager. This way FilterManager does't need to know anything about state managing, app state and global state and all this functionallity is contained inside FilterStateManager. It will also allow us to easier migrate to new stores, as there we won't need to change FilterManager at all then, but we will just update the FilterStateManager.

Reply:
I thought about this again.
The FilterStateManager just exposes an observable FilterManager subscribes to.
There is no reason for this interface to change.
If you insist I will do this, but I still feel like the other way around actually creates a stronger coupling.

Reply:
in the long run i think we want filterManager to be the source of truth right ?